Core/SqlStatementModel/ConstantCollectionExpression.cs

```diff
@@ -0,0 +1,109 @@
+// along with re-linq; if not, see http://www.gnu.org/licenses.
+//
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using Remotion.Utilities;
+
+namespace Remotion.Linq.SqlBackend.SqlStatementModel
+{
+  /// <summary>
+  /// Expression representing an <see cref="ICollection"/>, an <see cref="ICollection{T}"/> or an <see cref="IReadOnlyCollection{T}"/>.
+  /// </summary>
+  public class ConstantCollectionExpression : Expression
+  {
+    /// <summary>
+    /// The original expression.
+    /// </summary>
+    public ConstantExpression ConstantExpression { get; }
+
+    /// <summary>
+    /// Indicates whether the <see cref="ConstantCollectionExpression.ConstantExpression"/>'s
+    /// <see cref="System.Linq.Expressions.ConstantExpression.Value"/> is a collection without elements.
+    /// </summary>
+    public bool IsEmptyCollection { get; }
+    
+    /// <summary>
+    /// The <see cref="System.Type"/> of the <see cref="ConstantCollectionExpression.ConstantExpression"/>'s <see cref="ConstantExpression.Value"/>.
+    /// </summary>
+    public override Type Type { get; }
+
+    /// <summary>
+    /// Creates a <see cref="ConstantCollectionExpression"/> if the given <paramref name="constantExpression"/> is a collection.
+    /// </summary>
+    /// <exception cref="ArgumentException">
+    ///   If <paramref name="constantExpression"/> is <see langword="null"/>,
+    ///   its <see cref="ConstantExpression.Value"/> is <see langword="null"/>
+    ///   or the <see cref="ConstantExpression.Value"/> is not a collection.
+    /// </exception>
+    public ConstantCollectionExpression (ConstantExpression constantExpression)
+    {
+      ArgumentUtility.CheckNotNull ("constantExpression", constantExpression);
+
+      var value = constantExpression.Value;
+      if (!CollectionUtility.IsCollection (value))
+        throw new ArgumentException ("The given constant expression does not have a collection value.", nameof(constantExpression));
+
+      ConstantExpression = constantExpression;
+      Type = value.GetType();
+
+      var collection = (IEnumerable) ConstantExpression.Value;
+      IsEmptyCollection = !collection.Cast<object>().Any();
+    }
+
+    protected override Expression Accept (ExpressionVisitor visitor)
+    {
+      ArgumentUtility.CheckNotNull (nameof(visitor), visitor);
+
+      if (visitor is IConstantCollectionExpressionVisitor collectionExpressionVisitor)
+        return collectionExpressionVisitor.VisitConstantCollection (this);
+
+      return base.Accept (visitor);
+    }
+
+    protected override Expression VisitChildren (ExpressionVisitor visitor)
+    {
+      return this;
+    }
+
+    public override ExpressionType NodeType => ExpressionType.Constant;
+    
+    public override string ToString ()
+    {
+      return "(" + string.Join (",", ((IEnumerable)ConstantExpression.Value).OfType<object>().Select (o => o.ToString())) + ")";
```

Comment:
```
      return "(" + string.Join (",", ((IEnumerable)ConstantExpression.Value).Cast<object>().Select (o => o.ToString())) + ")";
```
OfType is an implict null-filter which is a good thing considering the use of obj.ToString(). However, it implies a conditional typecheck and the theoretical option of skipping items. Since String.Join would stringify objects anyway, there is no need to do an obj.ToString(). Therefor:

```suggestion
      return "(" + string.Join (",", (IEnumerable)ConstantExpression.Value) + ")";
```

We should consider that ToString for a looong list would result in a a lot of text so perhaps isntead we should do
```
      return "(" + string.Join (",", ((IEnumerable)ConstantExpression.Value).Cast<object>().Take(5)) + ",...)";
```

Reply:
Fancy! :)